fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/postgresql/PostgreSqlTranslator.java

```diff
@@ -0,0 +1,154 @@
+        return "23505".equals(x.getSQLState());
+    }
+
+    @Override
+    public boolean isAlreadyExists(SQLException x) {
+        return "42710".equals(x.getSQLState());
+    }
+
+    @Override
+    public boolean isLockTimeout(SQLException x) {
+        return false;
+    }
+
+    @Override
+    public boolean isDeadlock(SQLException x) {
+        final String sqlState = x.getSQLState();
+        return "40XL1".equals(sqlState) || "40XL2".equals(sqlState);
+    }
+
+    @Override
+    public boolean isConnectionError(SQLException x) {
+        String sqlState = x.getSQLState();
+        return sqlState != null && sqlState.startsWith("08");
+    }
+
+    @Override
+    public DataAccessException translate(SQLException x) {
+        if (isDeadlock(x)) {
+            return new LockException(x, true);
+        }
+        else if (isLockTimeout(x)) {
+            return new LockException(x, false);
+        }
+        else if (isConnectionError(x)) {
+            return new ConnectionException(x);
+        }
+        else if (isDuplicate(x)) {
+            return new UniqueConstraintViolationException(x);
+        }
+        else if (isUndefinedName(x)) {
+            return new UndefinedNameException(x);
+        }
+        else {
+            return new DataAccessException(x);
+        }
+    }
+
+    @Override
+    public boolean isUndefinedName(SQLException x) {
+        return "42X05".equals(x.getSQLState());
+    }
+
+    @Override
+    public void fillProperties(Properties p, ConnectionDetails cd) {
+        p.put("user", cd.getUser());
+        p.put("password", cd.getPassword());
+
+        if (cd.isSsl()) {
+            p.put("sslConnection", "true");
+        }
+
+        if (cd.isHA()) {
+            logger.warning("No HA support for PostgreSql");
+        }
+    }
+
+    @Override
+    public String timestampDiff(String left, String right, String alias) {
+        if (alias == null || alias.isEmpty()) {
+            return String.format("{fn timestampdiff(SQL_TSI_SECOND, %s, %s)}", left, right);
+        }
+        else {
+            return String.format("{fn timestampdiff(SQL_TSI_SECOND, %s, %s)} AS %s", left, right, alias);
+        }
+    }
+
+    @Override
+    public String createSequence(String name, int cache) {
+        // cache isn't supported by PostgreSql
+        return "CREATE SEQUENCE " + name;
+    }
+
+    @Override
+    public String reorgTableCommand(String tableName) {
+        // REORG TABLE not supported by PostgreSql
+        throw new UnsupportedOperationException("reorg table is not supported!");
+    }
+
+    @Override
+    public String getDriverClassName() {
+        return "org.postgresql.Driver";
+    }
+
+    @Override
+    public String getUrl(Properties connectionProperties) {
+        PostgreSqlPropertyAdapter adapter = new PostgreSqlPropertyAdapter(connectionProperties);
+
+        return "jdbc:postgresql://" + adapter.getHost() + ":" + adapter.getPort() + "/" + adapter.getDatabase();
```

Comment:
do we ever need to pass other properties in? How do we handle in Derby and Db2?

Reply:
similar in derby and db2, for db2, it uses more properties to support dashdb.

Reply:
OK, this was a curiosity question.  NO worries